src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs

```diff
@@ -4762,5 +4762,68 @@ static void verify(CSharpCompilation comp, string expectedAssemblyName)
                 Assert.Equal(expectedAssemblyName, comp.GetTypeByMetadataName("System.Runtime.CompilerServices.IsExternalInit").ContainingAssembly.Name);
             }
         }
+
+        [Fact, WorkItem(67079, "https://github.com/dotnet/roslyn/issues/67079")]
+        public void DoNotPickTypeFromSourceWithFileModifier()
+        {
+            var corlib_cs = """
+                namespace System
+                {
+                    public class Object { }
+                    public struct Int32 { }
+                    public struct Boolean { }
+                    public class String { }
+                    public class ValueType { }
+                    public struct Void { }
+                    public class Attribute { }
+                    public class AttributeUsageAttribute : Attribute
+                    {
+                        public AttributeUsageAttribute(AttributeTargets t) { }
+                        public bool AllowMultiple { get; set; }
+                        public bool Inherited { get; set; }
+                    }
+                    public struct Enum { }
+                    public enum AttributeTargets { }
+                }
+                """;
+
+            var source = """
+                namespace System.Runtime.CompilerServices
+                {
+                    file class IsExternalInit {}
+                }
+
+                public class C
+                {
+                    public string Property { get; init; }
+                }
+                """;
+
+            var corlibWithoutIsExternalInitRef = CreateEmptyCompilation(corlib_cs)
+                .EmitToImageReference();
+
+            var corlibWithIsExternalInitRef = CreateEmptyCompilation(corlib_cs + IsExternalInitTypeDefinition)
+                .EmitToImageReference();
+
+            {
+                // proper type in corlib and file type in source
+                var comp = CreateEmptyCompilation(source, references: new[] { corlibWithIsExternalInitRef });
```

Comment:
>corlibWithIsExternalInitRef [](http://example.com/codeflow?start=78&length=27)

It feels like we also should test with compilation reference

Reply:
There is a test helper, `CommonTestBase.AsReference(Compilation comp, bool useCompilationReference)`, which might allow the test to be parameterized with a `bool useCompilationReference` parameter.

Reply:
Had to set specific `EmitOptions` in order to avoid unrelated warning. Please verify, that such move is correct